docs/source/sft_trainer.mdx

```diff
@@ -0,0 +1,154 @@
+
+
+## Advanced usage
+
+### Packing dataset (`ConstantLengthDataset`)
+
+`SFTTrainer` supports the integration of packed dataset. This is done with the `ConstantLengthDataset` utility clss that returns constant length chunks of tokens from stream of text files. To enable the usage of this dataset class, simply pass `packing=True` to the `SFTTrainer` constructor.
+
+```python
+...
+
+trainer = SFTTrainer(
+    "facebook/opt-350m",
+    train_dataset=dataset,
+    dataset_text_field="text",
+    packing=True
+)
+
+trainer.train()
+```
+
+If your dataset has several fields that you want to combine, for example if the dataset has `question` and `answer` fields and you want to combine them, you can pass a formatting function to the trainer that will take care of that. For example:
+
+```python
+def formatting_func(example):
+    text = f"### Question: {example['question']}\n ### Answer: {example['answer']}"
+    return text
+
+trainer = SFTTrainer(
+    "facebook/opt-350m",
+    train_dataset=dataset,
+    dataset_text_field="text",
+    packing=True,
+    formatting_func=formatting_func
+)
+
+trainer.train()
+```
+You can also customize the `ConstantLengthDataset` much more by directly passing the arguments to the `SFTTrainer` constructor. Please refer to that class' signature for more information.
+
+### Control over the pretrained model
+
+You can directly pass the kwargs of `.from_pretrained` method to the `SFTTrainer`. For example, if you want to load a model in a different precision, analoguous to `model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m", torch_dtype=torch.bfloat16)`, you can do:
+
+```python
+...
+
+trainer = SFTTrainer(
+    "facebook/opt-350m",
+    train_dataset=dataset,
+    dataset_text_field="text",
+    torch_dtype=torch.bfloat16,
+)
+
+trainer.train()
+```
+Note that all key word arguments of `.from_pretrained` are supported. 
+
+### Training adapters
+
+We also support a tight integration with 🤗 PEFT library so that any user can conveniently train adapters and share them on the Hub instead of training the entire model
+
+```python
+from datasets import load_dataset
+from trl import SFTTrainer
+from peft import LoraConfig
+
+dataset = load_dataset("imdb", split="train")
+
+peft_config = LoraConfig(
+    r=16,
+    lora_alpha=32,
+    lora_dropout=0.05,
+    bias="none",
+    task_type="CAUSAL_LM",
+)
+
+trainer = SFTTrainer(
+    "EleutherAI/gpt-neo-125m",
+    train_dataset=dataset,
+    dataset_text_field="text",
+    torch_dtype=torch.bfloat16,
+    peft_config=peft_config
+)
+
+trainer.train()
+```
+
+Note that in case of training adapters, we manually add a saving callback to automatically save the adapters only:
+```python
+class PeftSavingCallback(TrainerCallback):
+    def on_save(self, args, state, control, **kwargs):
```

Comment:
Have you tested this in distributed settings? I found that one usually needs to ensure the model is saved only on the main process or you cannot reload the model due to a `meta device` error.

Reply:
Since the trainer is not integrated with `accelerate` I can't really test it out of the box in distrubuted setting it seems.. 
however looking at the trainer source code, it adds a lot of checks using `args.shoud_save`, I believe this automatically takes care of the distrubted setting: https://github.com/huggingface/transformers/blob/b8648290d2d97e7c7dbccd2d4a6a4f44e70d3b63/src/transformers/training_args.py#L1747